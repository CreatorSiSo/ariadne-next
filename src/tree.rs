use crate::Color;

// TODO Replace with From<T>?
pub trait Layout {
    fn layout(self) -> Element;
}

impl<T: std::fmt::Display> Layout for T {
    fn layout(self) -> Element {
        Element::Inline {
            text: self.to_string(),
            style: Style::default(),
        }
    }
}

// TODO Explain all kinds of elements
#[derive(Debug, Clone)]
pub enum Element {
    VStack {
        children: Vec<Element>,
        style: Style,
    },
    HStack {
        children: Vec<Element>,
        style: Style,
    },
    // Example of wrapping:
    // --------------------------------------------
    // | Inline | VStack | Box with width | Inlin |
    // |--------| VStack | an some wrappi |--------
    // |        | VStack | ng content     |       |
    // |        | VStack |-----------------       |
    // |        | VStack |                        |
    // |--------------------------                |
    // | e that doesnt quiet fit |                |
    // |--------------------------------------    |
    // | Elements that wont be broken apart: |    |
    // | - VStacks (like this one)           |    |
    // | - HStacks                           |    |
    // | - Boxes with width                  |    |
    // --------------------------------------------
    Box {
        children: Vec<Element>,
        /// Unicode width of the content
        /// - Some: Height is adjusted to fit the entire content
        /// - None: Height is always that of the tallest child
        width: Option<usize>,
        style: Style,
    },
    Inline {
        text: String,
        style: Style,
    },
}

impl Element {
    pub fn vstack(elements: impl IntoIterator<Item = Element>) -> Self {
        Self::VStack {
            children: elements.into_iter().collect(),
            style: Style::default(),
        }
    }

    pub fn hstack(elements: impl IntoIterator<Item = Element>) -> Self {
        Self::HStack {
            children: elements.into_iter().collect(),
            style: Style::default(),
        }
    }

    pub fn box_(elements: impl IntoIterator<Item = Element>, width: Option<usize>) -> Self {
        Self::Box {
            children: elements.into_iter().collect(),
            width,
            style: Style::default(),
        }
    }

    pub fn inline(text: impl ToString) -> Self {
        Self::Inline {
            text: text.to_string(),
            style: Style::default(),
        }
    }

    pub fn style(&self) -> &Style {
        match self {
            Element::VStack { style, .. } => style,
            Element::HStack { style, .. } => style,
            Element::Box { style, .. } => style,
            Element::Inline { style, .. } => style,
        }
    }
}

impl Styled<Element> for Element {
    fn with_style(self, style: Style) -> Element {
        match self {
            Element::VStack {
                children: elements, ..
            } => Element::VStack {
                children: elements,
                style,
            },
            Element::HStack {
                children: elements, ..
            } => Element::HStack {
                children: elements,
                style,
            },
            Element::Box {
                children: elements,
                width,
                ..
            } => Element::Box {
                children: elements,
                width,
                style,
            },
            Element::Inline { text, .. } => Element::Inline { text, style },
        }
    }
}

impl<T: Layout> Styled<Element> for T {
    fn with_style(self, style: Style) -> Element {
        self.layout().with_style(style)
    }
}

pub trait Styled<T> {
    fn with_style(self, style: Style) -> T;
}

#[derive(Debug, Default, Clone)]
pub struct Style {
    /// Color of the text
    pub fg_color: Option<Color>,
    /// Color of the background
    pub bg_color: Option<Color>,
    /// Additional formatting data
    flags: TextStyleFlags,
}

impl Style {
    pub fn new() -> Self {
        Self::default()
    }

    pub fn with_fg(mut self, color: Color) -> Self {
        self.fg_color = Some(color);
        self
    }

    pub fn with_bg(mut self, color: Color) -> Self {
        self.bg_color = Some(color);
        self
    }

    pub fn with_bold(mut self) -> Self {
        self.flags = match self.flags {
            TextStyleFlags::None | TextStyleFlags::Bold => TextStyleFlags::Bold,
            TextStyleFlags::Italic | TextStyleFlags::BoldItalic => TextStyleFlags::BoldItalic,
        };
        self
    }

    pub fn with_italic(mut self) -> Self {
        self.flags = match self.flags {
            TextStyleFlags::None | TextStyleFlags::Italic => TextStyleFlags::Italic,
            TextStyleFlags::Bold | TextStyleFlags::BoldItalic => TextStyleFlags::BoldItalic,
        };
        self
    }

    pub fn is_bold(&self) -> bool {
        matches!(
            self.flags,
            TextStyleFlags::Bold | TextStyleFlags::BoldItalic
        )
    }

    pub fn is_italic(&self) -> bool {
        matches!(
            self.flags,
            TextStyleFlags::Italic | TextStyleFlags::BoldItalic
        )
    }

    // TODO add set_bold and set_italic
}

#[derive(Debug, Default, Clone)]
// TODO How and what information should this store? Bitflags?
enum TextStyleFlags {
    #[default]
    None,
    Bold,
    Italic,
    BoldItalic,
}

pub enum BasicColor {
    Unset,
    Black,
    Red,
    Green,
    Yellow,
    Blue,
    Magenta,
    Cyan,
    White,
}
