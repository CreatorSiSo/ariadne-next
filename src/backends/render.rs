use crate::tree::{Element, Inline};
use itertools::Itertools;
use std::io;
use unicode_width::UnicodeWidthStr;

pub(super) trait Render {
    fn render(writer: &mut impl io::Write, element: Element) -> Result<(), io::Error> {
        let (_, height) = compute_size(&element);
        let mut lines = Vec::from_iter((0..height).map(|_| String::new()));

        Self::render_element(&mut lines, element);

        for line in lines {
            writeln!(writer, "{line}")?;
        }
        Ok(())
    }

    fn render_element(lines: &mut [String], element: Element) {
        match element {
            Element::VStack(elements) => Self::render_vstack(lines, elements),
            Element::HStack(elements) => Self::render_hstack(lines, elements),
            ref this @ Element::Box { ref content, .. } => Self::render_box(lines, this, content),
            Element::Inline(inline) => lines[0].push_str(&Self::render_inline(inline).to_string()),
        }
    }

    fn render_vstack(lines: &mut [String], elements: Vec<Element>) {
        let mut start = 0;
        for element in elements {
            let (_, height) = compute_size(&element);
            Self::render_element(&mut lines[start..start + height], element);
            start += height;
        }
    }

    fn render_hstack(lines: &mut [String], elements: Vec<Element>) {
        for element in elements {
            fill_spaces(lines);
            Self::render_element(lines, element);
        }
    }

    fn render_box(lines: &mut [String], element: &Element, content: &[Inline]) {
        let string: String = content.iter().map(|inline| inline.text.as_str()).collect();
        debug_assert!(!string.contains('\n'));

        fill_spaces(lines);
        let (width, _) = compute_size(element);
        for (index, chunk) in string.chars().chunks(width).into_iter().enumerate() {
            let line = &mut lines[index];
            for char in chunk {
                line.push(char);
            }
        }
    }

    fn render_inline(inline: Inline) -> impl std::fmt::Display;
}

fn fill_spaces(lines: &mut [String]) {
    let max_width = lines.iter().map(|line| line.width()).max().unwrap_or(0);
    for line in lines {
        line.push_str(&" ".repeat(max_width - line.width()));
    }
}

fn compute_size(element: &Element) -> (usize, usize) {
    match element {
        Element::VStack(stack) => stack
            .iter()
            .map(compute_size)
            .fold((0, 0), |(width, height), (w, h)| (width.max(w), height + h)),
        Element::HStack(stack) => stack
            .iter()
            .map(compute_size)
            .fold((0, 0), |(width, height), (w, h)| (width + w, height.max(h))),
        Element::Box { content, width, .. } => {
            let len: usize = content.iter().map(|inline| inline.text.len()).sum();
            if let Some(width) = width {
                (*width, len.div_ceil(*width))
            } else {
                (len, 1)
            }
        }
        Element::Inline(inline) => (
            inline.text.len(),
            1, /* TODO Set this to 0 when text is empty? */
        ),
    }
}

#[test]
fn test_rendering() {
    use crate::tree::{shortcuts::inline, Element, Inline, TextStyle};

    struct TestBackend;
    impl Render for TestBackend {
        fn render_inline(inline: Inline) -> impl std::fmt::Display {
            inline.text
        }
    }

    let element = Element::VStack(vec![
        inline("test1"),
        Element::HStack(vec![
            inline("1"),
            inline(" "),
            inline("2"),
            Element::Box {
                content: vec![Inline::new("#_#_#_#__#_#_#_##_#_#_#__#_#_#_#")],
                width: Some(8),
                style: TextStyle::default(),
            },
        ]),
        Element::Inline(Inline::new("test3")),
    ]);
    assert_eq!(compute_size(&element), (11, 6));

    let mut writer = Vec::new();
    TestBackend::render(&mut writer, element).unwrap();
    let output = String::from_utf8(writer).unwrap();

    // Expected:
    // test1
    // 1 2#_#_#_#_
    //    _#_#_#_#
    //    #_#_#_#_
    //    _#_#_#_#
    // test3
    insta::assert_snapshot!(output);
}
