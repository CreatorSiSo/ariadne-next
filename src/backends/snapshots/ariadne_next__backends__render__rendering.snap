---
source: src/backends/render.rs
expression: output
---
test1
1 2#_#_#_#_
   _#_#_#_#
   #_#_#_#_
   _#_#_#_#
test3

